src/Security/AuthenticationMiddleware.php

```diff
@@ -60,4 +135,136 @@ public function process(HTTPRequest $request, callable $delegate)
+        $defaultRedirectUrl = static::config()->get('expiredPasswordDefaultRedirect');
+
+        if ($sessionRedirectUrl || $defaultRedirectUrl) {
+            $redirectUrl = Director::absoluteURL($sessionRedirectUrl ?? $defaultRedirectUrl);
+        } else {
+            $redirectUrl = null;
+        }
+
+        if ($session && $redirectUrl) {
+            $whitelist = static::config()->get('expiredPasswordWhitelistedUrls');
+            if (!is_array($whitelist)) {
+                $whitelist = [];
+            }
+
+            $sessionWhitelist = $session->get(static::SESSION_KEY_EXPIRED_PASSWORD_WHITELIST);
+            if (is_array($sessionWhitelist)) {
+                $whitelist = array_merge($whitelist, $sessionWhitelist);
+            }
+
+            $currentUrl = Director::absoluteURL($request->getURL(true));
+
+            $allowedUrls = array_merge([
+                $redirectUrl
+            ], array_map(
+                function ($url) {
+                    return Director::absoluteURL($url);
+                },
+                $whitelist
+            ));
+
+            if (!in_array($currentUrl, $allowedUrls, true)) {
+                return $this->redirectOrForbid($request, $redirectUrl);
+            }
+        } else {
+            Security::setCurrentUser(null);
+        }
+    }
+
+    /**
+     * Returns a redirect to the URL if text/html is acceptable, otherwise
+     * deauthenticates the current request by Security::setCurrentUser(null)
+     *
+     * @param HTTPRequest $request
+     * @param string $redirectUrl
+     *
+     * @return HTTPResponse|null
+     */
+    private function redirectOrForbid(HTTPRequest $request, $redirectUrl)
+    {
+        $acceptableTypes = $request->getAcceptMimetypes();
+
+        $allowedTypes = [
+            '*/*',
+            'text/*',
+            'text/html',
+            'text/html',
+            'application/xhtml+xml',
+            'text/xml',
+            'application/xml'
+        ];
+
+        if (count(array_intersect($allowedTypes, $acceptableTypes)) > 0) {
+            $redirectAllowed = true;
+        } else {
+            //  if browser didn't send the Accept header
+            //  with mimetypes, let's redirect anyway
+            $redirectAllowed = count($acceptableTypes) === 0;
+        }
+
+        if ($redirectAllowed) {
+            $response = new HTTPResponse();
+            $response->redirect($redirectUrl);
+            return $response;
+        } else {
+            Security::setCurrentUser(null);
+        }
+    }
+
+    /**
+     * Preserve the password change URL in the session
+     * That URL is to be redirected to to force users change expired passwords
+     *
+     * @param Session $session Session where we persist the redirect URL
+     * @param string $url change password form address
+     */
+    public static function setExpiredPasswordRedirect(Session $session, $url)
```

Comment:
This doesn't appear to need to be static - you can presumably modify the state of the instance via `AuthenticationMiddleware::singleton()->setExpiredPasswordRedirect(...)`

Reply:
Do you mean we could call the static methods through instance that we get through DI? Otherwise, we're not really changing any AuthenticationMiddleware state in this method - that's why it's static.